junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1561,97 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
 		return result;
 	}
 
-	private static String formatIndexes(Deque<Integer> indexes) {
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed = 0;
+		while (expectedIterator.hasNext() && actualIterator.hasNext()) {
+			Object expectedElement = expectedIterator.next();
+			Object actualElement = actualIterator.next();
+
+			if (expectedElement == actualElement) {
+				processed++;
+				continue;
+			}
+
+			indexes.addLast(processed++);
```

Comment:
Okay. If I understood you correctly, does that mean you're looking for something more like this?

```java
int processed = 0;
while (expectedIterator.hasNext() && actualIterator.hasNext()) {
	processed++;
	Object expectedElement = expectedIterator.next();
	Object actualElement = actualIterator.next();

	if (expectedElement == actualElement) {
		continue;
	}

	indexes.addLast(processed - 1);
	assertIterableElementsEqual(expectedElement, actualElement, indexes, messageSupplier);
	indexes.removeLast();
}
```